### PR TITLE
feat (api, worker): workflow queue GRPC handlers

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/grpc/handlers.go [hunk 1]
```diff
@@ -29,3 +36,91 @@ func (h *handlers) AddBuildLog(stream BuildLog_AddBuildLogServer) error {
 		}
 	}
 }
+
+func (*handlers) SendLog(stream WorkflowQueue_SendLogServer) error {
```
Comment (yesnault) [inline engine/api/grpc/handlers.go]: doc

File: engine/api/grpc/handlers.go [hunk 2]
```diff
@@ -29,3 +36,91 @@ func (h *handlers) AddBuildLog(stream BuildLog_AddBuildLogServer) error {
...
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		log.Debug("grpc.SendLog> Got %+v", in)
+
+		db := database.GetDBMap()
+		if err := workflow.AddLog(db, nil, in); err != nil {
+			log.Warning("grpc.SendLog> Unable to insert log : %s", err)
```
Comment (yesnault) [inline engine/api/grpc/handlers.go]: wrap?

File: engine/api/grpc/handlers.go [hunk 3]
```diff
@@ -29,3 +36,91 @@ func (h *handlers) AddBuildLog(stream BuildLog_AddBuildLogServer) error {
...
+			return err
+		}
+	}
+}
+
+func (*handlers) SendResult(c context.Context, res *sdk.Result) (*empty.Empty, error) {
```
Comment (yesnault) [inline engine/api/grpc/handlers.go]: doc

File: engine/api/grpc/handlers.go [hunk 4]
```diff
@@ -29,3 +36,91 @@ func (h *handlers) AddBuildLog(stream BuildLog_AddBuildLogServer) error {
...
+	//Update spwan info
+	infos := []sdk.SpawnInfo{{
+		RemoteTime: remoteTime,
+		Message:    sdk.SpawnMsg{ID: sdk.MsgSpawnInfoWorkerEnd.ID, Args: []interface{}{workerName, res.Duration}},
+	}}
+
+	//Add spawn infos
+	if _, err := workflow.AddSpawnInfosNodeJobRun(tx, job.ID, infos); err != nil {
+		log.Error("addQueueResultHandler> Cannot save spawn info job %d: %s", job.ID, err)
+		return nil, err
```
Comment (yesnault) [inline engine/api/grpc/handlers.go]: wrap

File: engine/api/grpc/server.go [hunk 5]
```diff
@@ -41,40 +41,56 @@ func Init(port int, tls bool, certFile, keyFile string) error {
...
 
-func authorize(ctx context.Context) error {
+func authorize(ctx context.Context) (*sdk.Worker, error) {
 	if md, ok := metadata.FromContext(ctx); ok {
 		if len(md["name"]) > 0 && len(md["token"]) > 0 {
 			w, err := auth.GetWorker(database.GetDBMap(), md["token"][0])
 			if err != nil {
 				log.Error("grpc.authorize> Unable to get worker %v:%v => %s", md["name"], md["token"], err)
-				return sdk.ErrServiceUnavailable
+				return nil, sdk.ErrServiceUnavailable
```
Comment (yesnault) [inline engine/api/grpc/server.go]: wrap ?

File: engine/api/workflow/execute_node_job_run.go [hunk 6]
```diff
@@ -218,9 +218,11 @@ func BookNodeJobRun(id int64, hatchery *sdk.Hatchery) (*sdk.Hatchery, error) {
 
 //AddLog adds a build log
 func AddLog(db gorp.SqlExecutor, job *sdk.WorkflowNodeJobRun, logs *sdk.Log) error {
-	logs.PipelineBuildJobID = job.ID
+	if job != nil {
+		logs.PipelineBuildJobID = job.ID
 	logs.PipelineBuildID = job.WorkflowNodeRunID
-
+	}
```
Comment (sguiheux) [inline engine/api/workflow/execute_node_job_run.go]: gofmt?

File: engine/worker/cmd_register.go [hunk 7]
```diff
@@ -3,27 +3,37 @@ package main
 import (
 	"os"
 
+	"github.com/ovh/cds/engine/api/worker"
 	"github.com/ovh/cds/sdk/log"
 	"github.com/spf13/cobra"
```
Comment (sguiheux) [inline engine/worker/cmd_register.go]: import

File: sdk/cdsclient/interface.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package cdsclient
+
+import (
+	"context"
+
```
Comment (sguiheux) [inline sdk/cdsclient/interface.go]: no space

File: sdk/cdsclient/http.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+
```
Comment (sguiheux) [inline sdk/cdsclient/http.go]: no space

File: sdk/cdsclient/client.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
+package cdsclient
+
+import (
+	"io"
+	"net/http"
+	"time"
+
```
Comment (sguiheux) [inline sdk/cdsclient/client.go]: no space

File: engine/worker/cmd_main.go [hunk 11]
```diff
@@ -11,6 +11,8 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
 
+	"encoding/json"
```
Comment (sguiheux) [inline engine/worker/cmd_main.go]: wrong place

File: engine/worker/cmd_main.go [hunk 12]
```diff
@@ -89,84 +91,155 @@ func cmdMain(w *currentWorker) *cobra.Command {
...
+		go w.logger(ctx)
 
+		// start queue polling
 		pbjobs := make(chan sdk.PipelineBuildJob)
 		wjobs := make(chan sdk.WorkflowNodeJobRun)
 		errs := make(chan error)
 
-		go w.client.QueuePolling(ctx, wjobs, pbjobs, errs, 2*time.Second)
+		//Before start the loop, take the bookJobID
+		//TODO: TEST THIS !!
```
Comment (sguiheux) [inline engine/worker/cmd_main.go]: remove todo

File: engine/worker/run.go [hunk 13]
```diff
@@ -294,9 +291,13 @@ func workingDirectory(basedir string, jobInfo *worker.PipelineBuildJobInfo) stri
 	return dir
 }
 
-func (w *currentWorker) run(pbji *worker.PipelineBuildJobInfo) sdk.Result {
+func (w *currentWorker) run(ctx context.Context, pbji *worker.PipelineBuildJobInfo) sdk.Result {
 	t0 := time.Now()
 	defer func() { log.Info("Run Pipeline Build Job Done (%s)", sdk.Round(time.Since(t0), time.Second).String()) }()
+	/*
```
Comment (sguiheux) [inline engine/worker/run.go]: remove

File: sdk/keychain/keychain_linux_test.go [hunk 14]
```diff
@@ -1,11 +1,6 @@
 package keychain
 
-import (
-	"testing"
-
-	"github.com/stretchr/testify/assert"
-)
-
+/*
```
Comment (sguiheux) [inline sdk/keychain/keychain_linux_test.go]: remove

File: engine/worker/polling.go [hunk 15]
```diff
@@ -68,8 +68,23 @@ func (w *currentWorker) takeWorkflowJob(ctx context.Context, job sdk.WorkflowNod
 
 			}
 		}
-	}(cancel, pipelineBuildJobID)
+	}(cancel, job.ID)
+
+	// Reset build variables
+	w.currentJob.buildVariables = nil
+	//start := time.Now()
```
Comment (sguiheux) [inline engine/worker/polling.go]: remove

File: engine/api/main_routes.go [hunk 16]
```diff
@@ -199,6 +200,7 @@ func (router *Router) init() {
 	router.Handle("/queue/workflows/requirements/errors", NeedWorker(), POST(postWorkflowJobRequirementsErrorHandler))
 	router.Handle("/queue/workflows/{id}/take", NeedWorker(), POST(postTakeWorkflowJobHandler))
 	router.Handle("/queue/workflows/{id}/book", NeedHatchery(), POST(postBookWorkflowJobHandler))
+	router.Handle("/queue/workflows/{id}/infos", NeedWorker(), POST(getWorkflowJobHandler))
```
Comment (yesnault) [inline engine/api/main_routes.go]: GET

File: engine/worker/builtin_artifact.go [hunk 17]
```diff
@@ -156,33 +157,33 @@ func runArtifactDownload(a *sdk.Action, pbJob sdk.PipelineBuildJob, stepOrder in
...
 	}
 
-	sendLog(pbJob.ID, fmt.Sprintf("Downloading artifacts from %s-%s-%s/%s into '%s'...\n", project, application, pipeline, tag, filePath), pbJob.PipelineBuildID, stepOrder, false)
+	w.sendLog(pbJob.ID, fmt.Sprintf("Downloading artifacts from %s-%s-%s/%s into '%s'...\n", project, application, pipeline, tag, filePath), pbJob.PipelineBuildID, stepOrder, false)
 	err := sdk.DownloadArtifacts(project, application, pipeline, tag, filePath, environment)
 	if err != nil {
-		res.Status = sdk.StatusFail
+		res.Status = sdk.StatusFail.String()
 		res.Reason = fmt.Sprintf("%s\n", err)
 		log.Warning("Cannot download artifacts: %s\n", err)
```
Comment (yesnault) [inline engine/worker/builtin_artifact.go]: remove \n on warning

File: engine/worker/cmd_main.go [hunk 18]
```diff
@@ -85,73 +89,161 @@ func cmdMain() *cobra.Command {
...
+		pbjobs := make(chan sdk.PipelineBuildJob, 1)
+		wjobs := make(chan sdk.WorkflowNodeJobRun, 1)
+		errs := make(chan error, 1)
+
+		//Before start the loop, take the bookJobID
+		if w.bookedJobID != 0 {
+			log.Debug("Try to take the job %d", w.bookedJobID)
+			b, _, err := sdk.Request("GET", fmt.Sprintf("/queue/%d/infos", w.bookedJobID), nil)
+			if err != nil {
+				log.Error("Unable to load pipeline build job %d", w.bookedJobID)
```
Comment (yesnault) [inline engine/worker/cmd_main.go]: add err in msg

File: sdk/cdsclient/config.go [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package cdsclient
+
+type Config struct {
```
Comment (yesnault) [inline sdk/cdsclient/config.go]: doc

File: sdk/cdsclient/http.go [hunk 20]
```diff
@@ -0,0 +1,211 @@
+package cdsclient
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
```
Comment (yesnault) [inline sdk/cdsclient/http.go]: -> log on client

File: sdk/cdsclient/http.go [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+
+	var body []byte
+	body, err = ioutil.ReadAll(respBody)
+	if err != nil {
+		return nil, code, err
+	}
+
+	if c.config.Verbose {
+		if len(body) > 0 {
+			log.Printf("Response Body: %s\n", body)
```
Comment (yesnault) [inline sdk/cdsclient/http.go]: log on client

File: sdk/cdsclient/http.go [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+
+	return body, code, nil
+}
+
+// Stream makes an authenticated http request and return io.ReadCloser
+func (c *client) Stream(method string, path string, args []byte, mods ...RequestModifier) (io.ReadCloser, int, error) {
+	var savederror error
+
+	if c.config.Verbose {
+		log.Printf("Request %s Body : %s", c.config.Host+path, string(args))
```
Comment (yesnault) [inline sdk/cdsclient/http.go]: log on client

File: sdk/cdsclient/http.go (hunk 20 above)
Comment (fsamin) [inline sdk/cdsclient/http.go]: i'll do in a next PR

File: sdk/cdsclient/http.go (hunk 21 above)
Comment (fsamin) [inline sdk/cdsclient/http.go]: i'll do in a next PR

File: sdk/cdsclient/http.go (hunk 22 above)
Comment (fsamin) [inline sdk/cdsclient/http.go]: i'll do in a next PR

**Pull request merged**